first attempt at implementing compiler schema in Buildtest

shahzebsiddiqui opened the pull request:
This PR does the following
- Added class CompilerBuilder which is a subclass of BuilderBase that implements the compiler schema.
 - Add support for viewing compiler schema via 'buildtest show schema'
 - Modify global schema since we ran into validation error when specifying maintainer key. This was due to expected type.
 - Add compiler schema to buildtest with some additional key's including ``cxxflags``, ``fflags``, ``cppflags`` ``exec_args``

shahzebsiddiqui committed:
first attempt at implementing 'compiler' schema in Buildtest
Added class CompilerBuilder which is a subclass of BuilderBase that
implements the compiler schema.
Add support for viewing compiler schema via 'buildtest show schema'
Modify global schema since we ran into validation error when specifying
maintainer key. This was due to expected type.
Add compiler schema to buildtest.

shahzebsiddiqui committed:
Refactor code by adding lookup dictionary for language file extension
and compiler table. We add exception when any C, C++, Fortran compiler
is not set we raise an error. Replace if/else condition with more
simplified and reduced code structure. Add docstring to method
_build_testcontent
add pgi, cray to compiler schema.

shahzebsiddiqui committed:
Add Compiler class GNUCompiler, IntelCompiler, PGICompiler, CrayCompiler
that are subclass of CompilerBuilder which set cc, cxx, fc. We have a
get_compilers method to return the list of compiler wrapper. In
addition, we have get_path method which returns a key/value pair
of compiler wrapper and full path to binary. In future, we may extend
feature to invoke the compiler class by name of module to
create multiple instance of Compilers.
We removed detect_compiler method and refactored language detection
in method detect_lang

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+    # https://software.intel.com/content/www/us/en/develop/documentation/fortran-compiler-developer-guide-and-reference/top/compiler-setup/using-the-command-line/understanding-file-extensions.html
+    # Fortran Extensions: http://fortranwiki.org/fortran/show/File+extensions
+    lang_ext_table = {
+        ".c": "C",
```
These compiler-specific variables should belong with the appropriate class.

shahzebsiddiqui replied:
The file extension for C, C++, Fortran are compiler agnostic and not belong to any compiler class. You may be getting confused, the compiler class are not C, Fortran, C++ compiler but rather name of the compiler ``GNU``, ``Intel``, ``PGI``, etc.. Each of these compilers come with a C, C++, Fortran wrapper. That's why i kept ``lang_ext_table`` in CompilerBuilder, the main purpose is to use the extensions to detect the programming language (``self.lang``) which is also compiler agnostic.

vsoch replied:
Ah ok. Is this going to be a common practice - to have a huge table of extensions? Can you think of a simpler way to do this, maybe regular expressions that can be shared across functions and classes instead of a hard coded lookup just with the Compiler class?

shahzebsiddiqui replied:
Fortran is the culprit when it comes to many extensions, i converted the if/else into a lookup table which simplified the logic, the fact is you can write any C, C++ program without a file extension and it would still work so buildtest is trying to enforce some rules when user specify source file. 

Eventually i wanted to add ``.cu`` to the lookup table for CUDA language, it seems easier to add key, value so i can map extension to a language name. Another option was to have a separate list per language and then search in list. Either way it doesn't make much difference, I prefer a lookup table. 

```
c_ext = ['.c']
cplus_ext = ['.cpp', '.cc', ...]
fortran_ext = [.'f', ...]
```

vsoch replied:
Makes sense, let’s have this lookup table (since it’s shared between all compilers) stay as is for now, it will be inherited by child classes.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+        ".ftn": "Fortran",
+    }
+
+    cc = None
```
Same with these.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+        self.compiler_name = None
+        self.lang = self.detect_lang(self.source)
+
+        # Get GNU Compiler if 'gnu' defined in compiler properties.
```
These if else elif should be broken into the appropriate CompilerBuilder base - it doesn't make sense to have to repeat them all here. If there are shared operations before this step, create a common function (for the CompilerBuilder base) that is run/shared for all child classes that need it.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+            self.cc, self.cxx, self.fc = cray.get_compilers()
+
+        # Set ldflags, cppflags, cflags, cxxflags, fflags if defined in Buildspec
+        self.ldflags = self.compiler_recipe[self.compiler_name].get("ldflags")
```
This is also compiler specific.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+
+        ext = os.path.splitext(sourcefile)[1]
+
+        # if ext not in self.lang_ext_table then raise an error. This table consist of all file extensions that map to a Programming Language
```
You shouldn't need this complicated logic if the compilers are easily separated - the compiler will know a set of languages, and exit if there is mismatch.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+        """
+        cmd = []
+
+        # Generate C compilation line
```
Another example here of if else elif that will be addressed by separation into the appropriate classes. I shouldn't need to edit the Abstract Compiler Base if I want to tweak the FortranCompiler, for example. This is what I'd have to do now.

shahzebsiddiqui replied:
So the way C, C++, Fortran compilation works for single program is as follows, this is true for any compiler to the best of my knowledge.

```
$cc $CPPFLAGS $CFLAGS -o $EXEC $SOURCE $LDFLAGS
$cxx $CPPFLAGS $CXXFLAGS -o $EXEC $SOURCE $LDFLAGS
$fc $CPPFLAGS $FFLAGS -o $EXEC $SOURCE $LDFLAGS
```

The $cc is your C compiler with CFLAGS as option to C compiler, the same is true with Fortran (FFLAGS) and C++ (CXXFLAGS). 
That's how Makefiles work if you take look at this example, you just need to change cc and CFLAGS to map to the compiler you want but the pattern for compilation remains the same.
https://www.cs.swarthmore.edu/~newhall/unixhelp/howto_makefiles.html

That's why i implemented this in Base class. The one issue we have right now is CompilerBuilder is base class of BuilderBase in order to inherit the build logic.  The compilation pattern can change if user wants to build source into object files (.o) and link each object into executable. 

Here is some example https://www3.ntu.edu.sg/home/ehchua/programming/cpp/gcc_make.html for compilation

vsoch replied:
Thank you for your explanation of how compilers work - I'm aware :) This is different from the refactor needed to clean up the base class.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+        return list(filter(None, cmd))
+
+
+class GNUCompiler(CompilerBuilder):
```
This is a great start! This is exactly what I had in mind - see if you can simplify the CompilerBuilder base so that it doesn't have all that if else elif logic and you can just put shared functions there, and specifics in each of the classes here.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+           a list that contains content of the test.
+        """
+
+        self.compiler_recipe = self.recipe.get("compiler")
```
Is the compiler key guaranteed to be defined?

vsoch replied:
Same question for source.

shahzebsiddiqui replied:
both compiler and source are required see
https://github.com/shahzebsiddiqui/buildtest-1/blob/compiler_schema/buildtest/buildsystem/schemas/compiler/compiler-v0.0.1.schema.json#L6

https://github.com/shahzebsiddiqui/buildtest-1/blob/6fecf4c42212ab16aec9b18ba35d73c41cf5f3ac/buildtest/buildsystem/schemas/compiler/compiler-v0.0.1.schema.json#L33

The ``source`` key is required with oneOf the compiler names at the moment. For now this was what we discussed in #296

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+        # check for relative path.
+        self.source = resolve_path(sourcefile) or resolve_path(
+            os.path.join(
+                os.path.dirname(self.buildspec), self.compiler_recipe["source"]
```
If you defined `self.compiler_recipe["source"]` to go into `sourcefile` why don't you use `sourcefile` here as well?

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+        # Get GNU Compiler if 'gnu' defined in compiler properties.
+        if self.compiler_recipe.get("gnu"):
+            self.compiler_name = "gnu"
+            gnu = GNUCompiler()
```
Okay here is where I still think it needs work - right now you are passing _all_ kinds of compiler specs into the Compiler class, and doing this complicated if/then/else thing. The executors should be modular (and use different names as necessary) so we start, off the bat, with the recipe mapped to the right compiler. The Compiler class does not exist to be a jump off point to determine which other class to instantiate, but instead an Abstract class to implement common functions or variables. Take a look at snakemake.executors if you need a good example of modularity with parent/child classes.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+        self.fflags = self.compiler_recipe[self.compiler_name].get("fflags")
+        self.cxxflags = self.compiler_recipe[self.compiler_name].get("cxxflags")
+
+        cmd = self.generate_compile_cmd()
```
The rest of this recipe is hugely long, and should be broken into functions that can be mapped to steps. I would suggest taking these steps:

 - first refactor the logic to start with a compiler class, removing a lot of the code above with if / else because a Compiler class is just abstract to provide common functions.
 - then for each, determine what steps are common and can be shared. Write those into the Compiler parent class, and just call from the child classes. Leave the specific commands that are compiler specific.

The goal is that shared functionality is represented in neatly defined steps held by the parent Compiler Class. The children then call those steps, and it can be read from top to bottom to easily understand what is going on (it's not a big dump of spaghetti code as it is now).  Then of course each child will have some custom steps / functionality that is more prominent in the class.  I'll continue my review after this change.

shahzebsiddiqui committed:
unfinished work, major refactor for CompilerBuilder class to make
class more modular. Right now there are some issues with how test
content is generated.

shahzebsiddiqui committed:
move exec_args into compiler property as pose to compiler definition in schema.
2nd attempt at refactoring CompilerBuilder

shahzebsiddiqui committed:
add 'name' key in compiler to select the compiler type which
selects the appropriate Compiler class to invoke. Tested this
with intel and gnu compiler and it was successful.

shahzebsiddiqui committed:
invoked wrong Compiler class when compiler name is cray

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -640,3 +655,294 @@ def _build_testcontent(self):
+        """
+
+        if workdir:
+            return [f"{workdir}/{self.executable}"]
```
os.path.join here

shahzebsiddiqui replied:
good call. Ill fix this

vsoch reviewed (changes requested):
Looks good! One tiny path issue, then with the schema in this is a solid start.

shahzebsiddiqui committed:
use os.path.join instead of using fstring

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -787,7 +787,7 @@ def simple_run(self, workdir=None):
 
         if workdir:
-            return [f"{workdir}/{self.executable}"]
+            return [os.path.join(workdir / self.executable)]
```
Why not `os.path.join(workdir, self.executable)` ?

shahzebsiddiqui replied:
the reason why i put as a list is in future there could be different types of run cmd right now its a single line, as a standalone executable or running executable with arguments. Running test via mpi, or debugging and profiling code could be multiline commands that  are post build. It would be easier to add items to list, and the ``build_run_cmd`` could be expanded to encapsulate different run scenarios. There could be a combination of YAML and similar to ``run`` key from ``script`` to allow for more customized tests.

vsoch replied:
Sorry I wasn’t asking about the list - I’m asking about the /

shahzebsiddiqui replied:
My thought was to allow flexibility in future if user want's to run a test not relative to their working directory, they can provide a relative directory. For example, building a binary with different flags  needs to have a unique executable name, user could have way to organize where executable is generated and other artifacts as a result of execution of test.

Right now it's possible two sections in Buildspec that reference same sourcefile generate the same executable. This is because the default name of executable is related to source file. One possibility is to have a key for user to specify name of executable, but that is unrelated to your comment.

vsoch replied:
I’ll try again - why did you use a slash (/) instead of an expected comma for two args in is.path.join?

shahzebsiddiqui replied:
wow that's embarrassing this is not suppose to be a / it is suppose to be a comma, i will fix this.

shahzebsiddiqui committed:
remove self.compiler_setup() this method doesnt exist anymore.
Fix os.path.join in building run command

vsoch reviewed (approved):
If the schema is updated from the last PR, then I think this looks ready (for a first draft). I suspect you might change some things when you work more on it!
